### PR TITLE
Add pyamg demo

### PR DESCRIPTION
**Pull request by chrisrichardson**
A simple pyamg example.

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -35,7 +35,7 @@ dependencies = [
 [project.optional-dependencies]
 docs = ["markdown", "pyyaml", "sphinx", "sphinx_rtd_theme"]
 lint = ["ruff"]
-optional = ["numba"]
+optional = ["numba", "pyamg"]
```
Comment (garth-wells) [inline python/pyproject.toml:38]: `pyamg` shouldn't be added here, it's require for a demo but is not used anywhere in the core.

File: python/demo/demo_pyamg.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+g = ufl.sin(5 * x[0])
+a = form(inner(grad(u), grad(v)) * dx, dtype=dtype)
+L = form(inner(f, v) * dx + inner(g, v) * ds, dtype=dtype)
+
+A0 = assemble_matrix(a, [bc])
+b = assemble_vector(L)
+apply_lifting(b.array, [a], bcs=[[bc]])
+set_bc(b.array, [bc])
+
+A = csr_matrix((A0.data, A0.indices, A0.indptr))
```
Comment (garth-wells) [inline python/demo/demo_pyamg.py:84]: Use https://github.com/FEniCS/dolfinx/blob/1c3831c908d74c2ca29d94f90ce1cd28ee71c07b/python/dolfinx/la.py#L135

**Review by garth-wells — COMMENTED** [review]
This demo is more-or-less a copy of the existing Poisson demo. Could we combine, or maybe we could expand the demo to do more, e.g. add elasticity and used smoothed aggregation AMG?

File: python/pyproject.toml (hunk 1 above)
Comment (jorgensd) [inline python/pyproject.toml:38]:
```suggestion
optional = ["numba"]
demos = ["numba", "pyamg"]
```

Comment (chrisrichardson) [reply python/pyproject.toml:38]: I'm just removing it for now.

**Comment by chrisrichardson** [comment]
> This demo is more-or-less a copy of the existing Poisson demo. Could we combine, or maybe we could expand the demo to do more, e.g. add elasticity and used smoothed aggregation AMG?

Yes, now included. Probably needs a little reorganisation - maybe two functions that are called from main?

File: python/pyproject.toml (hunk 1 above)
Comment (jorgensd) [reply python/pyproject.toml]: Removing it means that it will never be executed on CI?

**Comment by garth-wells** [comment]
> > This demo is more-or-less a copy of the existing Poisson demo. Could we combine, or maybe we could expand the demo to do more, e.g. add elasticity and used smoothed aggregation AMG?
> 
> Yes, now included. Probably needs a little reorganisation - maybe two functions that are called from main?

Re-factoring into functions would be good, see related #3076.

**Pull request merged**